compiler/src/dotty/tools/dotc/typer/ProtoTypes.scala

```diff
@@ -350,23 +355,15 @@ object ProtoTypes {
 
     override def withContext(newCtx: Context) =
       if (newCtx `eq` ctx) this
-      else {
-        val result = new FunProto(args, resType, typer)(newCtx)
-        result.myTypedArgs = myTypedArgs
-        result.myTypedArg = myTypedArg
-        result.evalState = evalState
-        result.myTupled = myTupled
-        result.toDrop = toDrop
-        result
-      }
+      else new FunProto(args, resType)(typer, state)(newCtx)
```

Comment:
> The idea of cloning FunProtos has promise, but we need to make sure that all clones refer to the same core state.

In what situation are multiple clones of the same FunProto live at the same time, and why is it necessary that they share the same state?

Reply:
I noted we had problems with auto-tupling otherwise. The other danger is that we have FP1, infer a type argument, then go to FP2, infer another argument, then decide this is a blocker so fall back to FP1 and try something else afterwards. In this case the computed type argument of FP2 is lost.